### PR TITLE
Feature/remove hardcoded

### PR DESCRIPTION
**Pull request by vincentcreusotfx**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,6 +71,7 @@ The webhook listen on port 9877 by default.
 The webhook config is done in `config/servicenow.yml`.
 
 All `default_incident` properties supports Go templating with structure defined in [AlertManager documentation](https://prometheus.io/docs/alerting/notifications/#data).
+Every key value pair in `default_incident` is sent as attributes of the incident created.
```
Comment (louisfelix) [inline README.md]:
```suggestion
Every key value pair in `default_incident` is sent as an incident field.
```

File: main.go [hunk 2]
```diff
@@ -344,7 +320,7 @@ func applyTemplate(name string, text string, data template.Data) (string, error)
 	return result.String(), nil
 }
 
-func validateIncident(incident Incident) error {
+func validateIncident(incident Incident) {
 	impact := incident["impact"].(string)
```
Comment (louisfelix) [inline main.go]: You must now check if `incident["impact"]` exist before the type assertion to string

Comment (vincentcreusotfx) [reply main.go]: Modified and added test

File: main_test.go [hunk 3]
```diff
@@ -374,3 +377,21 @@ TOTO
...
+	incident["urgency"] = "toto"
+	validateIncident(incident)
+	incident["impact"] = ""
+	incident["urgency"] = ""
+	validateIncident(incident)
+	incident["impact"] = nil
+	incident["urgency"] = nil
+	validateIncident(incident)
+	// test should not panic
+}
```
Comment (rolandvarga) [inline main_test.go]: I think these should all be moved into a testing table

File: main.go [hunk 4]
```diff
@@ -344,20 +320,17 @@ func applyTemplate(name string, text string, data template.Data) (string, error)
 	return result.String(), nil
 }
 
-func validateIncident(incident Incident) error {
-	impact := incident["impact"].(string)
-	if len(impact) > 0 {
-		if _, err := strconv.Atoi(impact); err != nil {
+func validateIncident(incident Incident) {
```
Comment (rolandvarga) [inline main.go]: I think this should return an error anyways wether we proceed with incident creation or not

File: main.go [hunk 5]
```diff
@@ -344,20 +320,17 @@ func applyTemplate(name string, text string, data template.Data) (string, error)
...
+
+	if impact, ok := incident["impact"]; ok && impact != nil && len(impact.(string)) > 0 {
+		if _, err := strconv.Atoi(impact.(string)); err != nil {
```
Comment (rolandvarga) [inline main.go]: If it's possible I would use either `strings(var)` or `var.(string)` everywhere, but not both

Comment (vincentcreusotfx) [reply main.go]: Actually I'm always var.(string)ing no ?

Comment (louisfelix) [reply main.go]: Both "type assertion" and "cast" can be used depending on the situation (depending if we work with an interface or a concrete type), so, no, I don't think it makes sense to "choose" one ([ref](https://stackoverflow.com/questions/20494229/what-is-the-difference-between-type-conversion-and-type-assertion))

Comment (rolandvarga) [reply main.go]: okay, sorry I didn't realize you actually need to assert here. Ignore my comment!

File: main_test.go [hunk 6]
```diff
@@ -289,7 +289,7 @@ func TestApplyTemplate_emptyText(t *testing.T) {
 		t.Fatal(err)
 	}
 	expected := ""
-	if result != expected {
+	if string(result) != expected {
 		t.Errorf("Unexpected result: got %v, want %v", result, expected)
```
[the remark this answers is not in the document]
Comment (vincentcreusotfx) [reply main_test.go]: yep changing

File: main.go [hunk 7]
```diff
@@ -272,27 +257,27 @@ func onResolvedGroup(data template.Data, updatableIncident Incident) error {
...
 
+	errs := applyIncidentTemplate(incident, data)
+	logErrors(errs)
+	errs = validateIncident(incident)
+	logErrors(errs)
 	return incident, nil
 }
 
+func logErrors(errs []error) {
+	for err := range errs {
```
Comment (louisfelix) [inline main.go]: This is not working, should be something like:
```suggestion
	for _, err := range errs {
```

Comment (rolandvarga) [reply main.go]: This isn't a comment on functionality but more like a personal opinion: I think we should return a single error in all of these cases and use the idiomatic GO error handling pattern instead. I think it's more readable that way.

File: main.go [hunk 8]
```diff
@@ -345,19 +326,24 @@ func applyTemplate(name string, text string, data template.Data) (string, error)
...
-	impact := incident["impact"].(string)
-	if len(impact) > 0 {
-		if _, err := strconv.Atoi(impact); err != nil {
-			log.Errorf("'impact' field value is '%v' but should be an integer, please fix your configuration. Incident creation/update will proceed but this field will be missing", impact)
+	var str strings.Builder
+	if impact, ok := incident["impact"]; ok && impact != nil && len(impact.(string)) > 0 {
+		if _, err := strconv.Atoi(impact.(string)); err != nil {
+			str.WriteString("'impact' field value is ")
+			str.WriteString(impact.(string))
+			str.WriteString("but should be an integer, please fix your configuration. Incident creation/update will proceed but this field will be missing. ")
```
Comment (louisfelix) [inline main.go]:
```suggestion
			str.WriteString(" but should be an integer, please fix your configuration. Incident creation/update will proceed but this field will be missing. ")
```

File: main.go [hunk 9]
```diff
@@ -345,19 +326,24 @@ func applyTemplate(name string, text string, data template.Data) (string, error)
...
 
-	urgency := incident["urgency"].(string)
-	if len(urgency) > 0 {
-		if _, err := strconv.Atoi(urgency); err != nil {
-			log.Errorf("'urgency' field value is '%v' but should be an integer, please fix your configuration. Incident creation/update will proceed but this field will be missing", urgency)
+	if urgency, ok := incident["urgency"]; ok && urgency != nil && len(urgency.(string)) > 0 {
+		if _, err := strconv.Atoi(urgency.(string)); err != nil {
+			str.WriteString("'urgency' field value is ")
+			str.WriteString(urgency.(string))
+			str.WriteString("but should be an integer, please fix your configuration. Incident creation/update will proceed but this field will be missing.")
```
Comment (louisfelix) [inline main.go]:
```suggestion
			str.WriteString(" but should be an integer, please fix your configuration. Incident creation/update will proceed but this field will be missing.")
```

**Pull request merged**